CVR export re-upload fails

On `master` with a fresh DB. Ballot manifest re-upload _does_ work.

To repro, upload a ballot manifest, upload a CVR export, then _re_-upload the same CVR export.

Server error:
```
01:36:08,492  INFO corla:377 - endpoint /import-cvr-export hit by localhost:8888
01:36:08,501  INFO corla:377 - endpoint /contest/county hit by localhost:8888
01:36:08,503  INFO corla:423 - successful /contest/county by countyadmin1 from localhost:8888
01:36:08,531  INFO AbstractBatchImpl:193 - HHH000010: On release of batch it still contained JDBC statements
01:36:08,531 ERROR BatchingBatch:120 - HHH000315: Exception executing batch [java.sql.BatchUpdateException: Batch entry 0 insert into contest (county_id, description, name, version, votes_allowed, id) values (1, '', 'Regent of the University of Colorado - At Large', 0, 1, 392),(1, '', 'COUNTY COMMISSIONER DISTRICT 3', 0, 1, 394) was aborted: ERROR: duplicate key value violates unique constraint "ukdv45ptogm326acwp45hm46uaf"
  Detail: Key (name, county_id, description, votes_allowed)=(Regent of the University of Colorado - At Large, 1, , 1) already exists.  Call getNextException to see other errors in the batch.], SQL: insert into contest (county_id, description, name, version, votes_allowed, id) values (?, ?, ?, ?, ?, ?)
01:36:08,531  WARN SqlExceptionHelper:129 - SQL Error: 0, SQLState: 23505
01:36:08,531 ERROR SqlExceptionHelper:131 - Batch entry 0 insert into contest (county_id, description, name, version, votes_allowed, id) values (1, '', 'Regent of the University of Colorado - At Large', 0, 1, 392),(1, '', 'COUNTY COMMISSIONER DISTRICT 3', 0, 1, 394) was aborted: ERROR: duplicate key value violates unique constraint "ukdv45ptogm326acwp45hm46uaf"
  Detail: Key (name, county_id, description, votes_allowed)=(Regent of the University of Colorado - At Large, 1, , 1) already exists.  Call getNextException to see other errors in the batch.
01:36:08,531  WARN SqlExceptionHelper:129 - SQL Error: 0, SQLState: 23505
01:36:08,531 ERROR SqlExceptionHelper:131 - ERROR: duplicate key value violates unique constraint "ukdv45ptogm326acwp45hm46uaf"
  Detail: Key (name, county_id, description, votes_allowed)=(Regent of the University of Colorado - At Large, 1, , 1) already exists.
javax.persistence.PersistenceException: org.hibernate.exception.ConstraintViolationException: could not execute batch
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:147)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:155)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1423)
	at org.hibernate.query.Query.getResultList(Query.java:146)
	at org.hibernate.query.criteria.internal.compile.CriteriaQueryTypeQueryAdapter.getResultList(CriteriaQueryTypeQueryAdapter.java:72)
	at us.freeandfair.corla.query.CountyContestResultQueries.matching(CountyContestResultQueries.java:77)
	at us.freeandfair.corla.csv.DominionCVRExportParser.addContests(DominionCVRExportParser.java:244)
	at us.freeandfair.corla.csv.DominionCVRExportParser.parse(DominionCVRExportParser.java:372)
	at us.freeandfair.corla.endpoint.CVRExportImport.parseFile(CVRExportImport.java:114)
	at us.freeandfair.corla.endpoint.CVRExportImport.endpoint(CVRExportImport.java:163)
	at us.freeandfair.corla.Main.lambda$activateEndpoints$5(Main.java:253)
	at spark.RouteImpl$1.handle(RouteImpl.java:72)
	at spark.http.matching.Routes.execute(Routes.java:61)
	at spark.http.matching.MatcherFilter.doFilter(MatcherFilter.java:130)
	at spark.embeddedserver.jetty.JettyHandler.doHandle(JettyHandler.java:50)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1568)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:564)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:317)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:110)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.Invocable.invokePreferred(Invocable.java:128)
	at org.eclipse.jetty.util.thread.Invocable$InvocableExecutor.invoke(Invocable.java:222)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:294)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:199)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:672)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:590)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.hibernate.exception.ConstraintViolationException: could not execute batch
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:112)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:111)
	at org.hibernate.engine.jdbc.batch.internal.BatchingBatch.performExecution(BatchingBatch.java:121)
	at org.hibernate.engine.jdbc.batch.internal.BatchingBatch.doExecuteBatch(BatchingBatch.java:97)
	at org.hibernate.engine.jdbc.batch.internal.AbstractBatchImpl.execute(AbstractBatchImpl.java:147)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.getBatch(JdbcCoordinatorImpl.java:195)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:2983)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3503)
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:89)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:589)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:463)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:337)
	at org.hibernate.event.internal.DefaultAutoFlushEventListener.onAutoFlush(DefaultAutoFlushEventListener.java:50)
	at org.hibernate.internal.SessionImpl.autoFlushIfRequired(SessionImpl.java:1396)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1481)
	at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1445)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1414)
	... 28 more
Caused by: java.sql.BatchUpdateException: Batch entry 0 insert into contest (county_id, description, name, version, votes_allowed, id) values (1, '', 'Regent of the University of Colorado - At Large', 0, 1, 392),(1, '', 'COUNTY COMMISSIONER DISTRICT 3', 0, 1, 394) was aborted: ERROR: duplicate key value violates unique constraint "ukdv45ptogm326acwp45hm46uaf"
  Detail: Key (name, county_id, description, votes_allowed)=(Regent of the University of Colorado - At Large, 1, , 1) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:148)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2191)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:472)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:791)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1563)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeBatch(NewProxyPreparedStatement.java:2544)
	at org.hibernate.engine.jdbc.batch.internal.BatchingBatch.performExecution(BatchingBatch.java:111)
	... 42 more
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "ukdv45ptogm326acwp45hm46uaf"
  Detail: Key (name, county_id, description, votes_allowed)=(Regent of the University of Colorado - At Large, 1, , 1) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2477)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2190)
	... 47 more
01:36:08,532 ERROR corla:89 - could not query database for contest results
01:36:08,532  INFO corla:131 - could not parse malformed CVR export file 386: java.lang.IllegalStateException: no running transaction
01:36:08,534 ERROR corla:427 - error 422 bad data from client on /import-cvr-export: malformed CVR export file 386 by countyadmin1 from localhost:8888
```

--- a/server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java
+++ b/server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java
@@ -39,7 +39,6 @@ import us.freeandfair.corla.model.UploadedFile.FileStatus;
 import us.freeandfair.corla.model.UploadedFile.HashStatus;
 import us.freeandfair.corla.persistence.Persistence;
 import us.freeandfair.corla.query.CastVoteRecordQueries;
-import us.freeandfair.corla.query.ContestQueries;
 import us.freeandfair.corla.query.CountyContestResultQueries;
 
 /**

--- a/server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java
+++ b/server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java
@@ -39,6 +39,7 @@ import us.freeandfair.corla.model.UploadedFile.FileStatus;
 import us.freeandfair.corla.model.UploadedFile.HashStatus;
 import us.freeandfair.corla.persistence.Persistence;
 import us.freeandfair.corla.query.CastVoteRecordQueries;
+import us.freeandfair.corla.query.ContestQueries;
 
 /**
  * The "CVR export import" endpoint.
@@ -111,6 +112,7 @@ public class CVRExportImport extends AbstractCountyDashboardEndpoint {
                                       Persistence.getByID(the_file.countyID(), County.class));
       final int deleted = 
           CastVoteRecordQueries.deleteMatching(the_file.countyID(), RecordType.UPLOADED);
+      ContestQueries.deleteForCounty(the_file.countyID());
       if (parser.parse()) {
         final int imported = parser.recordCount().getAsInt();
         Main.LOGGER.info(imported + " CVRs parsed from file " + the_file.id());

--- a/server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java
+++ b/server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java
@@ -40,6 +40,7 @@ import us.freeandfair.corla.model.UploadedFile.HashStatus;
 import us.freeandfair.corla.persistence.Persistence;
 import us.freeandfair.corla.query.CastVoteRecordQueries;
 import us.freeandfair.corla.query.ContestQueries;
+import us.freeandfair.corla.query.CountyContestResultQueries;
 
 /**
  * The "CVR export import" endpoint.
@@ -112,7 +113,7 @@ public class CVRExportImport extends AbstractCountyDashboardEndpoint {
                                       Persistence.getByID(the_file.countyID(), County.class));
       final int deleted = 
           CastVoteRecordQueries.deleteMatching(the_file.countyID(), RecordType.UPLOADED);
-      ContestQueries.deleteForCounty(the_file.countyID());
+      CountyContestResultQueries.deleteForCounty(the_file.countyID());
       if (parser.parse()) {
         final int imported = parser.recordCount().getAsInt();
         Main.LOGGER.info(imported + " CVRs parsed from file " + the_file.id());

--- a/server/eclipse-project/src/main/java/us/freeandfair/corla/query/ContestQueries.java
+++ b/server/eclipse-project/src/main/java/us/freeandfair/corla/query/ContestQueries.java
@@ -100,4 +100,19 @@ public final class ContestQueries {
 
     return result;
   }
+  
+  /**
+   * Deletes all the contests for the county with the specified ID.
+   * 
+   * @param the_id The county ID.
+   */
+  public static void deleteForCounty(final Long the_county_id) {
+    final Set<Contest> contests = 
+        forCounty(Persistence.getByID(the_county_id, County.class));
+    if (contests != null) {
+      for (final Contest c : contests) {
+        Persistence.delete(c);
+      }
+    }
+  }
 }

--- a/server/eclipse-project/src/main/java/us/freeandfair/corla/query/ContestQueries.java
+++ b/server/eclipse-project/src/main/java/us/freeandfair/corla/query/ContestQueries.java
@@ -114,5 +114,6 @@ public final class ContestQueries {
         Persistence.delete(c);
       }
     }
+    Persistence.flush();
   }
 }

--- a/server/eclipse-project/src/main/java/us/freeandfair/corla/query/CountyContestResultQueries.java
+++ b/server/eclipse-project/src/main/java/us/freeandfair/corla/query/CountyContestResultQueries.java
@@ -129,4 +129,21 @@ public final class CountyContestResultQueries {
 
     return result;
   }
+  
+  /**
+   * Deletes all the contest results for the county with the specified ID.
+   * 
+   * @param the_id The county ID.
+   */
+  public static void deleteForCounty(final Long the_county_id) {
+    final Set<CountyContestResult> results = 
+        forCounty(Persistence.getByID(the_county_id, County.class));
+    if (results != null) {
+      for (final CountyContestResult c : results) {
+        Persistence.delete(c);
+        Persistence.delete(c.contest());
+      }
+    }
+    Persistence.flush();
+  }
 }

